app/libs/installations/crashlytics/api.rb

```diff
@@ -0,0 +1,193 @@
+        Installations::Response::Keys.transform([crash_data], transforms).first
+      end
+    end
+
+    def list_apps(transforms)
+      execute { firebase_management_service.list_apps(transforms) }
+    end
+
+    private
+
+    # Fetch and merge analytics and crashlytics data
+    def fetch_crash_data(app_id, app_version)
+      analytics = release_data(app_id)[:analytics_data].find { |a| a[:version_name] == app_version } || {}
+      crashlytics = release_data(app_id)[:crashlytics_data].find { |c| c[:version_name] == app_version } || {}
+      analytics.merge(crashlytics)
+    end
+
+    # Ensure required keys are present in the crash data
+    def ensure_required_keys(crash_data, transforms)
+      transforms.each do |old_key, new_key|
+        crash_data[new_key] ||= crash_data[old_key] if crash_data.has_key?(old_key)
+      end
+      crash_data
+    end
+
+    # Memoized BigQuery client
+    def bigquery_client
+      @bigquery_client ||= BIGQUERY.new(credentials: key_file)
+    end
+
+    # Query dataset names and return relevant datasets
+    def datasets
+      execute do
+        target_datasets = {}
+        bigquery_client.datasets.each do |dataset|
+          case dataset.dataset_id
+          when /^analytics_\d+$/ then target_datasets[:ga4] = dataset_pattern(dataset)
+          when /crashlytics/i then target_datasets[:crashlytics] = dataset_pattern(dataset)
+          else
+            Rails.logger.warn("Dataset ID does not match expected datasets: #{dataset.dataset_id}")
+          end
+        end
+        target_datasets
+      end
+    end
+
+    # Build the dataset pattern for query purposes
+    def dataset_pattern(dataset)
+      "#{dataset.project_id}.#{dataset.dataset_id}.*"
+    end
+
+    # Firebase APIs
+    def firebase_management_service
+      Installations::Google::Firebase::Api.new(project_number, json_key)
+    end
+
+    def firebase_client
+      execute { firebase_management_service.set_firebase_client }
+    end
+
+    # Query string for Crashlytics data
+    def crashlytics_query(dataset_name)
+      execute { build_crashlytics_query(dataset_name) }
+    end
+
+    # Query string for Analytics data
+    def analytics_query(dataset_name, app_id)
+      execute { build_analytics_query(dataset_name, app_id) }
+    end
+
+    # Run the BigQuery query and fetch the data
+    def get_data(query)
+      bigquery_client.query(query)
+    end
+
+    # Fetch release data (memoized to avoid unnecessary calls)
+    def release_data(app_id)
+      @release_data ||= begin
+        analytics_data = get_data(analytics_query(datasets[:ga4], app_id))
+        crashlytics_data = get_data(crashlytics_query(datasets[:crashlytics]))
+        {analytics_data: analytics_data, crashlytics_data: crashlytics_data}
+      end
+    end
+
+    # Build the SQL query for Crashlytics data
+    def build_crashlytics_query(dataset_name)
+      <<-SQL.squish
+        WITH combined_events AS (
+          SELECT * FROM `#{dataset_name}`
```

Comment:
**fix**: we should limit the amount of data that we process in BigQuery as much as possible since this dataset for an actual app might be too big for the query. additionally, BigQuery queries are priced by the amount of data processed in the query.

From [BigQuery documentation](https://cloud.google.com/bigquery/pricing#on_demand_pricing):
> BigQuery uses a columnar data structure. You're charged according to the total data processed in the columns you select, and the total data per column is calculated based on the types of data in the column.

I would suggest adding a time-based filter to look at only the last 15 days of data.

secondly, we should add a filter on the app's bundle identifier since the dataset can contain data across multiple apps.

thirdly, we should avoid doing `select *` and just selecting the columns that we need.

the `combined_data` query may look something like this:

```
WITH combined_events AS (
          SELECT colum1, colum2, column3 FROM `#{dataset_name}`
          WHERE event_timestamp >= TIMESTAMP_SUB(CURRENT_TIMESTAMP(), INTERVAL 15 DAY)
          AND bundle_identifier = #{app_bundle_identifier}
        )
```